State machine, Board UI, and Frequency Modulation Sine PWM

anish-sivakumar opened the pull request:
- Updated FSM to have a stopping state
- Changed potState to uint16_t
- Added code to run the motor at speed up to 22Hz/50RPM depending on POT input 
- Program flows from INIT -> STARTUP -> RUNNING (if SW1 pressed) -> STOPPING (if SW1 pressed again) -> RUNNING (if SW1 pressed again)
- Validated switching between states in the lab 
- Validated higher speed motor operation with parallel DC Power Supply configuration and constant current/current limiting

anish-sivakumar committed:
Calculate offsets and read phase A,B currents

anish-sivakumar committed:
hall signal direction change

anish-sivakumar committed:
ADC current processing, STARTUP state structure, and LPF WIP

chyggen committed:
added board UI motor enable button

chyggen committed:
added potentiometer to board UI

anish-sivakumar committed:
Merge branch 'mc-board-ui' of https://github.com/anish-sivakumar/eBike into adc-dev

anish-sivakumar committed:
Run and Stop motor states with SW1

anish-sivakumar committed:
Frequency modulation with POT and button state changes

chyggen commented on motor_controller/motorbench-test-project/motorbench-test-project/rb_library/rb_isr.h:
```diff
@@ -74,34 +75,66 @@ void __attribute__((interrupt, auto_psv)) HAL_ADC_ISR(void)
             RB_FocInit(&PMSM);
             
             RB_FixedFrequencySinePWMInit(); //for testing
+            RB_BoardUIInit(&boardUI);
```
I think this is my own error but this should probably be in the mainInit function. On that note, forgive me if you already did this but we should look to only put stuff in here that needs to happen when the motor is starting back up from fault/stop.

If you don't think this is a small task then lets merge this and clean it up later.

anish-sivakumar replied:
How will the `boardUI` object be shared if it's used in `MainInit()` and in the ISR? I agree it should be in in `MainInit()`

chyggen commented on motor_controller/motorbench-test-project/motorbench-test-project/rb_library/rb_isr.h:
```diff
@@ -74,34 +75,66 @@ void __attribute__((interrupt, auto_psv)) HAL_ADC_ISR(void)
-            // RPM = 120*f/52
-            RB_FixedFrequencySinePWM(9);  // 9,10 tested
-            
+            RB_ADCRead(&PMSM.currentCal, &PMSM.iabc, &PMSM.vDC);
```
Is this just for observing on the scope? No need to change anything, just curious

anish-sivakumar replied:
`PMSM.iabc` will hold the 3-phase currents that are used as feedback inputs into the FOC controller. Reading them is the first step of FOC.

chyggen commented on motor_controller/motorbench-test-project/motorbench-test-project/rb_library/rb_isr.h:
```diff
@@ -20,31 +20,31 @@ extern "C" {
-
+RB_BOARD_UI boardUI;
+uint16_t speedLevel = 9;
 
```
Doesn't have to be delt with rn but we should move the hall global variable to here for consistency

anish-sivakumar replied:
Agreed, I will do this now

chyggen commented on motor_controller/motorbench-test-project/motorbench-test-project/rb_library/rb_measure.c:
```diff
@@ -1,6 +1,6 @@
+#include "adc/adc1.h"
 
 
 void RB_ADCCompensationInit(RB_MEASURE_CURRENT_T *pcal)
```
Slightly confusing naming. Is cal short for calibration? If so, what is the difference between calibration and compensation? If you decide to keep the names the same, maybe add a description of the function and parameters in the .h file

anish-sivakumar replied:
cal is short for calibration.
Calibration is determining the offsets
Compensation is applying the offsets
pcal holds the calibration/compensation constants.

I added comments in the .h

chyggen replied:
Ok that makes sense, the adc compensation function requires the original calibration settings. But doesn't that mean this is the  _calibration_ init and the RB_MeasureCurrentOffsetStepISR is the adc calibration step?

anish-sivakumar replied:
yeah

chyggen commented on motor_controller/motorbench-test-project/motorbench-test-project/rb_library/rb_measure.c:
```diff
@@ -20,63 +20,59 @@ void RB_ADCCompensationInit(RB_MEASURE_CURRENT_T *pcal)
 *
 */
-void RB_MeasureCurrentOffset(RB_MEASURE_CURRENT_T *pcal)
+void RB_MeasureCurrentOffsetStepISR(RB_MEASURE_CURRENT_T *pcal)
```
Same as above

chyggen commented on motor_controller/motorbench-test-project/motorbench-test-project/rb_library/rb_measure.h:
```diff
@@ -99,24 +99,53 @@ typedef struct
 void RB_ADCRead(RB_MEASURE_CURRENT_T *pcal, MC_ABC_T *piabc, int16_t *pvDC);
 
-inline static int16_t RB_ADCProcess(int16_t measurement, int16_t offset)
+inline static int16_t RB_ADCProcess(int16_t measurement, int16_t offset, int16_t gain)
```
Is there a reason for this to be inline in the .h, not in the .c?

anish-sivakumar replied:
it's inline cause motorBench had it as inline.

I had it in the .h because it's called in the .c file by a different function,  and we would have to make sure it was at the top of the file. (defined before usage)

chyggen commented on motor_controller/motorbench-test-project/motorbench-test-project/rb_library/rb_measure.h:
```diff
@@ -99,24 +99,53 @@ typedef struct
     
-    int16_t                 calibrationCounter;  /** counted number of samples used to calc offset */
-    int16_t                 calibrationComplete; /** 1 when offset values have been calculated */
+    int16_t                 calCounter;  /** counted number of samples used to calc offset */
```
is cal for calculation or calibration? maybe change to calc/calib to be more explicit

anish-sivakumar replied:
i meant calibration, good idea

chyggen commented on motor_controller/motorbench-test-project/motorbench-test-project/rb_library/rb_measure.h:
```diff
@@ -99,24 +99,53 @@ typedef struct
+ * coefficient for LPF given by f3db*Ts*2*pi, 
+ * where f3db is cutoff frequency and Ts is sampling time
+ */
+inline static int16_t RB_LPF(int16_t input, int16_t prevOutput, int16_t coeff)
```
Same as above: Is there a reason for this to be inline in the .h, not in the .c?

anish-sivakumar replied:
Will be used in the .c file, and would need to be at top of .c file

chyggen commented on motor_controller/motorbench-test-project/motorbench-test-project/rb_library/rb_isr.h:
```diff
@@ -74,34 +75,66 @@ void __attribute__((interrupt, auto_psv)) HAL_ADC_ISR(void)
-            RB_FixedFrequencySinePWM(9);  // 9,10 tested
-            
+            RB_ADCRead(&PMSM.currentCal, &PMSM.iabc, &PMSM.vDC);
             RB_HALL_Estimate();
```
Should pass hall struct once global is moved

chyggen reviewed (changes requested):
Great code style overall, this will be easy for someone in the future to come in and understand what's going on. Aside from the few comments on the "cal" naming, all my comments are just trying to understand some design choices and general thoughts for moving forward.

anish-sivakumar committed:
Requested changes, Found hall period issue at 22Hz sine

anish-sivakumar committed:
ADC function naming changes

anish-sivakumar commented:
- passing in hall object
- added comments to ADC calibration/compensation functions and renamed
- Still need to map current amperage to ADC reading (will do tomorrow)
- Need to test out LPF after (tomorrow)
 
- I need help figuring out the issue with hall.period at 22Hz Sine frequency. Can you check discord?

chyggen committed:
only run hall ISR routine if sector changes
